File: DSA-C++/5-sorting/4-merge-sort.cpp
#include <bits/stdc++.h>

using namespace std;


// Merge sort - O(n.logn) , O(n)
void merge (vector <int> &arr, int low, int mid, int high) {
    vector <int> temp;
    int left = low;
    int right = mid +1;

    while (left <= mid && right <= high) {
        if(arr[left] > arr[right]) {
            temp.push_back(arr[right]);
            right++;
        } 
        else {
            temp.push_back(arr[left]);
            left++;
        }
    }

    // push leftover elements
    while (right <= high) {
        temp.push_back(arr[right]);
        right++;
    }

    while (left <= mid) {
        temp.push_back(arr[left]);
        left++;
    }

    // update original array
    for (auto i=low; i <= high; i++) {
        arr[i] = temp[i-low];
    }
}

void divide(vector <int> &arr, int low, int high) {
    if (low >= high) return;

    int mid = (low + high) / 2;

    divide(arr, mid+1, high);
    divide(arr, low, mid);

    merge(arr, low, mid, high);
}


int main() {
    vector <int> arr = {78,34,38,2,3,5,9};
    
    divide(arr, 0, arr.size() - 1);

    for (auto x:arr) cout << x << " "; 
    
    cout << endl;
    return 0;
}

File: DSA-Python/4-sorting/4-merge-sort.py


# Worst case time complexity - O(n . log(n))
# Avg. case time complexity - O(n . log(n))
# Best case time complexity - O(n . log(n))
# Space complexity - O(n)
def merge(low:int, mid:int, high:int, nums:list) -> None:
    left, right = low, mid+1
    temp = []
    
    while left <= mid and right <= high:
        if nums[left] <= nums[right]:
            temp.append(nums[left])
            left += 1
        else:
            temp.append(nums[right])
            right += 1
            
    while right <= high:
        temp.append(nums[right])
        right += 1
        
    while left <= mid:
        temp.append(nums[left])
        left += 1
        
    for i in range(low, high+1):
        nums[i] = temp[i-low]
         

def divide(low: int, high:int, nums:list) -> None:
    if low >= high:
        return
    
    mid = (low + high) // 2
    
    divide(low, mid, nums)
    divide(mid+1, high, nums)
    
    merge(low, mid, high, nums)
    
    
if __name__ == "__main__":
    loops = int(input())
    
    for _ in range(loops):
        nums = list(map(int, input().split()))

        divide(0, len(nums)-1, nums)
        
        print(*nums)
        
        
        
'''
5
3 12 2 6 24 43
-2 45 54 3 12 23
8 8 8 8 8 8 0
0 -1 -4 -4 -4 -4 0
1, 2, 3, 4, 5
'''